### PR TITLE
Introduce BackgroundLayerChooser

### PR DESCRIPTION
**Pull request by KaiVolland**
## Description

This introduces the `BackgroundLayerChooser` and its childcomponent `BackgroundLayerPreview`.
It`s a combination of an overview map and a layer chooser with an immediate preview.

react-geo-backgroundlayer-chooser

## Pull request type

- [ ] Bugfix
- [x] Feature
- [ ] Dependency updates
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

## Do you introduce a breaking change?

- [ ] Yes
- [x] No

## Checklist

- [x] I understand and agree that the changes in this PR will be licensed under the
  [BSD-2-Clause](https://github.com/terrestris/react-geo/blob/main/LICENSE).
- [x] I have followed the [guidelines for contributing](https://github.com/terrestris/react-geo/blob/main/CONTRIBUTING.md).
- [x] The proposed change fits to the content of the [Code of Conduct](https://github.com/terrestris/react-geo/blob/main/CODE_OF_CONDUCT.md).
- [ ] I have added or updated tests and documentation, and the test suite passes (run `npm test` locally).
- [x] I have added a screenshot/screencast to illustrate the visual output of my update.

### CODE REVIEW

**Review by annarieger — APPROVED** [review]

File: src/BackroundLayerPreview/BackgroundLayerPreview.tsx [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  }, [zoom, center]);
+
+  const getBgLayersFromMap = (): OlLayer[] => {
+    return mainMap?.getLayerGroup().getLayers()
+      .getArray().filter(backgroundLayerFilter) as OlLayer[] || [];
+  };
+
+  const updateBgLayerVisibility = (evt: React.MouseEvent<HTMLDivElement>) => {
+    const target = evt?.currentTarget;
+    const layerId = target.dataset.uid;
```
Comment (LukasLohoff) [inline src/BackroundLayerPreview/BackgroundLayerPreview.tsx:90]:
```suggestion
    const layerId = target?.dataset?.uid;
```

File: src/BackgroundLayerChooser/BackgroundLayerChooser.example.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
+The BackgroundlayerChooser
```
Comment (LukasLohoff) [inline src/BackgroundLayerChooser/BackgroundLayerChooser.example.md:1]:
```suggestion
The BackgroundLayerChooser
```

File: src/BackroundLayerPreview/BackgroundLayerPreview.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    layers: previewLayer && [previewLayer]
+  }));
+
+  const [loading, setLoading] = useState<boolean>(false);
+
+  useEffect(() => {
+    // @ts-ignore
+    previewMap.on('loadstart', () => setLoading(true));
+    // @ts-ignore
+    previewMap.on('loadend', () => setLoading(false));
```
Comment (LukasLohoff) [inline src/BackroundLayerPreview/BackgroundLayerPreview.tsx:71]: don't you also need a cleanup here?

File: src/BackgroundLayerChooser/BackgroundLayerChooser.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+import OlOverviewMap from 'ol/control/OverviewMap';
+import OlLayerBase from 'ol/layer/Base';
+import OlLayer from 'ol/layer/Layer';
+import OlView from 'ol/View';
+import OlLayerTile from 'ol/layer/Tile';
+import OlLayerImage from 'ol/layer/Image';
+import { ObjectEvent } from 'ol/Object';
+import { getUid } from 'ol/util';
+
+import BackgroundLayerPreview from '../BackroundLayerPreview/BackgroundLayerPreview';
```
Comment (LukasLohoff) [inline src/BackgroundLayerChooser/BackgroundLayerChooser.tsx:12]:
```suggestion
import BackgroundLayerPreview from '../BackgroundLayerPreview/BackgroundLayerPreview';
```
there's a typo in the directory name (**Backround**LayerPreview)

File: src/index.ts [hunk 5]
```diff
@@ -50,11 +50,15 @@ import UserChip from './UserChip/UserChip';
 import Window from './Window/Window';
 import SearchResultsPanel from './Panel/SearchResultsPanel/SearchResultsPanel';
 import ClickAwayListener from './Util/ClickAwayListener/ClickAwayListener';
+import BackgroundLayerChooser from './BackgroundLayerChooser/BackgroundLayerChooser';
+import BackgroundLayerPreview from './BackroundLayerPreview/BackgroundLayerPreview';
```
Comment (LukasLohoff) [inline src/index.ts:54]:
```suggestion
import BackgroundLayerPreview from './BackgroundLayerPreview/BackgroundLayerPreview';
```

**Review by LukasLohoff — APPROVED** [review]
Very nice, can't wait to use this!

Left some suggestions on a few minor things, see comments below / above.

File: src/BackgroundLayerChooser/BackgroundLayerChooser.tsx [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        ovLayer = new OlLayerTile({
+          source: selectedLayer.getSource()
+        });
+      } else if (selectedLayer instanceof OlLayerImage){
+        ovLayer = new OlLayerImage({
+          source: selectedLayer.getSource()
+        });
+      }
+      if (ovLayer) {
+        const target = document.getElementById('overview-map');
```
Comment (dnlkoch) [inline src/BackgroundLayerChooser/BackgroundLayerChooser.tsx:87]: Absolutely minor and untested, but we might want to use a [`ref`](https://reactjs.org/docs/refs-and-the-dom.html) to the div element instead.

File: src/BackgroundLayerChooser/BackgroundLayerChooser.example.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+const layers = [
+  new OlLayerTile({
+    name: 'OSM',
+    source: new OlSourceOsm(),
+    isBackgroundLayer: true
+  }),
+  new OlLayerTile({
+    name: 'BKG',
+    isBackgroundLayer: true,
```
Comment (dnlkoch) [inline src/BackgroundLayerChooser/BackgroundLayerChooser.example.md:25]:
```suggestion
    properties: {
      name: 'BKG',
      isBackgroundLayer: true
    }
```

File: src/BackgroundLayerChooser/BackgroundLayerChooser.example.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import BackgroundLayerChooser from '@terrestris/react-geo/BackgroundLayerChooser/BackgroundLayerChooser';
+import MapComponent from '@terrestris/react-geo/Map/MapComponent/MapComponent';
+import MapContext from '@terrestris/react-geo/Context/MapContext/MapContext';
+import { useMap } from '@terrestris/react-geo/Hook/useMap';
+
+const layers = [
+  new OlLayerTile({
+    name: 'OSM',
+    source: new OlSourceOsm(),
+    isBackgroundLayer: true
```
Comment (dnlkoch) [inline src/BackgroundLayerChooser/BackgroundLayerChooser.example.md:21]:
```suggestion
    source: new OlSourceOsm(),
    properties: {
      name: 'OSM',
      isBackgroundLayer: true
    }
```

File: src/BackroundLayerPreview/BackgroundLayerPreview.tsx [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+      className={`layer-preview${isActive ? ' selected' : ''}`}
+      key={uid}
+      data-uid={uid}
+      onMouseOver={updateBgLayerVisibility}
+      onMouseLeave={restoreBgLayerVisibility}
+      onClick={updateBgLayerVisibility}
+    >
+      <Spin spinning={loading}>
+        <MapComponent
+          mapDivId={'previewmap-' + uid}
```
Comment (dnlkoch) [inline src/BackroundLayerPreview/BackgroundLayerPreview.tsx:132]:
```suggestion
          mapDivId={`previewmap-${uid}`}
```

File: src/BackgroundLayerChooser/BackgroundLayerChooser.tsx [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+        .find(c => c instanceof OlOverviewMap);
+      if (existingControl) {
+        map.removeControl(existingControl);
+      }
+      let ovLayer;
+      if (selectedLayer instanceof OlLayerTile) {
+        ovLayer = new OlLayerTile({
+          source: selectedLayer.getSource()
+        });
+      } else if (selectedLayer instanceof OlLayerImage){
```
Comment (dnlkoch) [inline src/BackgroundLayerChooser/BackgroundLayerChooser.tsx:81]:
```suggestion
      } else if (selectedLayer instanceof OlLayerImage) {
```

File: src/BackgroundLayerChooser/BackgroundLayerChooser.tsx [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+  };
+
+  return (
+    <div className={'bg-layer-chooser'}>
+      {
+        selectedLayer &&
+        <div className="layer-cards" style={{
+          maxWidth: layerOptionsVisible ? '100vw' : 0,
+          opacity: layerOptionsVisible ? 1 : 0
+        }}>
```
Comment (dnlkoch) [inline src/BackgroundLayerChooser/BackgroundLayerChooser.tsx:116]:
```suggestion
        <div 
          className="layer-cards" 
          style={{
            maxWidth: layerOptionsVisible ? '100vw' : 0,
            opacity: layerOptionsVisible ? 1 : 0
          }}>
```

**Review by dnlkoch — APPROVED** [review]
Very nice, thanks!

**Pull request merged**